src/number_theory/function_field.lean

```diff
@@ -107,4 +109,119 @@ is_integral_closure.is_dedekind_domain Fq[X] (ratfunc Fq) F _
 
 end ring_of_integers
 
+/-! ### The place at infinity on Fq(t) -/
+
+section infty_valuation
+
+variable [ dec : decidable_eq (ratfunc Fq)]
+include dec
+/-- The valuation at infinity is the nonarchimedean valuation on `Fq(t)` with uniformizer `1/t`.
+Explicitly, if `f/g ∈ Fq(t)` is a nonzero quotient of polynomials, its valuation at infinity is
+`multiplicative.of_add(degree(f) - degree(g))`. -/
+def infty_valuation_def (r : ratfunc Fq) : with_zero (multiplicative ℤ) :=
+if r = 0 then 0 else (multiplicative.of_add ((r.num.nat_degree : ℤ) - r.denom.nat_degree))
+
+lemma infty_valuation.map_zero' : infty_valuation_def Fq 0 = 0 := if_pos rfl
+
+lemma infty_valuation.map_one' : infty_valuation_def Fq 1 = 1 :=
+(if_neg one_ne_zero).trans $
+  by simp only [polynomial.nat_degree_one, ratfunc.num_one, int.coe_nat_zero, sub_zero,
+    ratfunc.denom_one, of_add_zero, with_zero.coe_one]
+
+lemma infty_valuation.map_mul' (x y : ratfunc Fq) :
+  infty_valuation_def Fq (x * y) = infty_valuation_def Fq x * infty_valuation_def Fq y :=
+begin
+  rw [infty_valuation_def, infty_valuation_def, infty_valuation_def],
+  by_cases hx : x = 0,
+  { rw [hx, zero_mul, if_pos (eq.refl _), zero_mul] },
+  { by_cases hy : y = 0,
+    { rw [hy, mul_zero, if_pos (eq.refl _), mul_zero] },
+    { rw [if_neg hx, if_neg hy, if_neg (mul_ne_zero hx hy), ← with_zero.coe_mul,
+        with_zero.coe_inj, ← of_add_add],
+      apply congr_arg,
+      rw [add_sub, sub_add, sub_sub_assoc_swap, sub_sub, sub_eq_sub_iff_add_eq_add],
+      norm_cast,
+      rw [← polynomial.nat_degree_mul x.denom_ne_zero y.denom_ne_zero,
+        ← polynomial.nat_degree_mul (ratfunc.num_ne_zero (mul_ne_zero hx hy))
+          (mul_ne_zero x.denom_ne_zero y.denom_ne_zero),
+        ← polynomial.nat_degree_mul (ratfunc.num_ne_zero hx) (ratfunc.num_ne_zero hy),
+        ← polynomial.nat_degree_mul (mul_ne_zero (ratfunc.num_ne_zero hx) (ratfunc.num_ne_zero hy))
+          (x * y).denom_ne_zero, ratfunc.num_denom_mul]} }
+end
+
+omit dec
+
+variable {Fq}
+/-- Equivalent fractions have the same valuation. -/
+lemma infty_valuation_well_defined {r₁ r₂ s₁ s₂ : polynomial Fq} (hr₁ : r₁ ≠ 0) (hs₁ : s₁ ≠ 0)
+  (hr₂ : r₂ ≠ 0) (hs₂ : s₂ ≠ 0) (h_eq : r₁*s₂ = r₂*s₁) :
+  (r₁.nat_degree : ℤ) - s₁.nat_degree = (r₂.nat_degree : ℤ) - s₂.nat_degree :=
+begin
+  rw sub_eq_sub_iff_add_eq_add,
+  norm_cast,
+  rw [← polynomial.nat_degree_mul hr₁ hs₂, ← polynomial.nat_degree_mul hr₂ hs₁, h_eq]
+end
+
+lemma ratfunc.num_add_ne_zero {x y : ratfunc Fq} (hxy : x + y ≠ 0) :
+  x.num * y.denom + x.denom * y.num ≠ 0 :=
+begin
+  intro h_zero,
+  have h := ratfunc.num_denom_add x y,
+  rw [h_zero, zero_mul] at h,
+  exact (mul_ne_zero (ratfunc.num_ne_zero hxy) (mul_ne_zero x.denom_ne_zero y.denom_ne_zero)) h
+end
```

Comment:
These should hold in more generality, I think in `R(x)` for any domain `R`. Could you move them to an appropriate place in `field_theory/ratfunc.lean`?

Reply:
It seems that `num` and `denom` are only defined if `R` is a field. I can move these lemmas to the `num_denom` section of `ratfunc.lean`, but I should probably change the names if I do this, do you have any suggestions?

(I'll do this once the latest commit is built so I can download the oleans, right now I keep having memory errors, even after building this file locally).

Reply:
I've moved these lemmas to `ratfunc.lean` and renamed some of them; if you think there are more appropriate names I would be happy to change them.